002-es6-modules.md

```diff
@@ -0,0 +1,246 @@
+# ES6 Modules
+
+Status | Draft
+------ | ----
+Author | Bradley Meck
+Date   | 7 Jan 2016
+
+It is our intent to:
+
+* implement interoperability for ES6 modules and node's existing module system
+* create a **Registry** compatible with the WhatWG Loader Registry
+
+## Purpose
+
+1. Allow a common module syntax for Browser and Server.
+2. Allow a common registry for inspection by Browser and Server environments/tools.
+    * these will most likely be represented by metaproperties like `import.ctx` but spec is not in place fully yet.
+
+## Related
+
+---
+
+### [ECMA262](https://tc39.github.io/ecma262/#sec-source-text-module-records)
+
+---
+
+Discusses the syntax and semantics of related syntax, and introduces:
+    
+#### Types
+    
+* [ModuleRecord](https://tc39.github.io/ecma262/#sec-abstract-module-records)
+    - Defines the list of imports via [ImportEntry](https://tc39.github.io/ecma262/#table-39).
+    - Defines the list of exports via [ExportEntry](https://tc39.github.io/ecma262/#table-41).
+    
+* [ModuleNamespace](https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects)
+    - Represents a read-only snapshot of a module's exports.
+    
+### Operations
+    
+* [ParseModule](https://tc39.github.io/ecma262/#sec-parsemodule)
+    - Creates a ModuleRecord from source code.
+    
+* [HostResolveImportedModule](https://tc39.github.io/ecma262/#sec-hostresolveimportedmodule)
+    - A hook for when an import is exactly performed.
+    
+* [CreateImportBinding](https://tc39.github.io/ecma262/#sec-createimportbinding)
```

Comment:
I don't think we should inject those via an import binding since we want to give the source text all the variables that will be in scope, currently we bootstrap the scope via an IIFE, but was thinking `var` more apt. (can't use `let` outside strict mode for now).

But we are also not talking 1-1 with the spec at the advisory v8 API at the bottom. If we want to make a 1-1 with the spec, it is much harder to integrate against v8.